tx/textual/valuerenderer/coins.go

```diff
@@ -0,0 +1,171 @@
+func NewCoinsValueRenderer(q CoinMetadataQueryFn) ValueRenderer {
+	return coinsValueRenderer{q}
+}
+
+type coinsValueRenderer struct {
+	// coinMetadataQuerier defines a function to query the coin metadata from
+	// state.
+	coinMetadataQuerier CoinMetadataQueryFn
+}
+
+var _ ValueRenderer = coinsValueRenderer{}
+
+func (vr coinsValueRenderer) Format(ctx context.Context, v protoreflect.Value, w io.Writer) error {
+	if vr.coinMetadataQuerier == nil {
+		return fmt.Errorf("expected non-nil coin metadata querier")
+	}
+
+	// Check whether we have a Coin or some Coins.
+	switch protoCoins := v.Interface().(type) {
+	// If it's a repeated Coin:
+	case protoreflect.List:
+		{
+			coins, metadatas := make([]*basev1beta1.Coin, protoCoins.Len()), make([]*bankv1beta1.Metadata, protoCoins.Len())
+			var err error
+			for i := 0; i < protoCoins.Len(); i++ {
+				coin := protoCoins.Get(i).Interface().(protoreflect.Message).Interface().(*basev1beta1.Coin)
+				coins[i] = coin
+				metadatas[i], err = vr.coinMetadataQuerier(ctx, coin.Denom)
+				if err != nil {
+					return err
+				}
+			}
+
+			formatted, err := formatCoins(coins, metadatas)
+			if err != nil {
+				return err
+			}
+
+			_, err = w.Write([]byte(formatted))
+			return err
+		}
+	// If it's a single Coin:
+	case protoreflect.Message:
+		{
+			coin := v.Interface().(protoreflect.Message).Interface().(*basev1beta1.Coin)
+
+			metadata, err := vr.coinMetadataQuerier(ctx, coin.Denom)
+			if err != nil {
+				return err
+			}
+
+			formatted, err := formatCoin(coin, metadata)
+			if err != nil {
+				return err
+			}
+
+			_, err = w.Write([]byte(formatted))
+			return err
+		}
+	default:
+		return fmt.Errorf("got invalid type %t for coins", v.Interface())
+	}
+
+}
+
+func (vr coinsValueRenderer) Parse(_ context.Context, r io.Reader) (protoreflect.Value, error) {
+	panic("implement me")
+}
+
+// formatCoin formats a sdk.Coin into a value-rendered string, using the
+// given metadata about the denom. It returns the formatted coin string, the
+// display denom, and an optional error.
+func formatCoin(coin *basev1beta1.Coin, metadata *bankv1beta1.Metadata) (string, error) {
+	coinDenom := coin.Denom
+
+	// Return early if no display denom or display denom is the current coin denom.
+	if metadata == nil || metadata.Display == "" || coinDenom == metadata.Display {
+		vr, err := formatDecimal(coin.Amount)
+		return vr + " " + coin.Denom, err
+	}
+
+	dispDenom := metadata.Display
+
+	// Find exponents of both denoms.
+	var coinExp, dispExp uint32
+	foundCoinExp, foundDispExp := false, false
+	for _, unit := range metadata.DenomUnits {
+		if coinDenom == unit.Denom {
+			coinExp = unit.Exponent
+			foundCoinExp = true
+		}
+		if dispDenom == unit.Denom {
+			dispExp = unit.Exponent
+			foundDispExp = true
+		}
+	}
+
+	// If we didn't find either exponent, then we return early.
+	if !foundCoinExp || !foundDispExp {
+		vr, err := formatInteger(coin.Amount)
+		return vr + " " + coin.Denom, err
```

Comment:
Must ensure that `coin.Denom` is printable ASCII.

Reply:
Related: https://github.com/cosmos/cosmos-sdk/issues/12026

I added the issue to the current SDK sprint. Once we merge that, then a simple comment in this code linking to it is IMO enough. Though can still add a simple check here for extra safety if you think it's useful.

Reply:
Re: https://github.com/cosmos/cosmos-sdk/pull/12729/files#r963716452, this is another area where we may need a slightly different implementation from `sdk.Coin` and/or need to change the way denom validation works generally to ensure bijectivity.